cornac/models/amr/recom_amr.py

```diff
@@ -0,0 +1,298 @@
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+# ============================================================================
+
+import numpy as np
+from tqdm.auto import tqdm
+
+from ..recommender import Recommender
+from ...exception import CornacException
+from ...exception import ScoreException
+from ...utils import fast_dot
+from ...utils.common import intersects
+from ...utils import get_rng
+from ...utils.init_utils import zeros, xavier_uniform
+
+
+class AMR(Recommender):
+    """Adversarial Training Towards Robust Multimedia Recommender System.
+
+    Parameters
+    ----------
+    k: int, optional, default: 10
+        The dimension of the gamma latent factors.
+
+    k2: int, optional, default: 10
+        The dimension of the theta latent factors.
+
+    n_epochs: int, optional, default: 20
+        Maximum number of epochs for SGD.
+
+    batch_size: int, optional, default: 100
+        The batch size for SGD.
+
+    learning_rate: float, optional, default: 0.001
+        The learning rate for SGD.
+
+    lambda_w: float, optional, default: 0.01
+        The regularization hyper-parameter for latent factor weights.
+
+    lambda_b: float, optional, default: 0.01
+        The regularization hyper-parameter for biases.
+
+    lambda_e: float, optional, default: 0.0
+        The regularization hyper-parameter for embedding matrix E and beta prime vector.
+
+    lmd: float, optional, default: 1.0
+        The regularization hyper-parameter in Eq. (27) for the adversarial sample loss.
+
+    use_gpu: boolean, optional, default: True
+        Whether or not to use GPU to speed up training.
+
+    trainable: boolean, optional, default: True
+        When False, the model is not trained and Cornac assumes that the model already \
+        pre-trained (U and V are not None).
+
+    verbose: boolean, optional, default: True
+        When True, running logs are displayed.
+
+    init_params: dictionary, optional, default: None
+        Initial parameters, e.g., init_params = {'Bi': beta_item, 'Gu': gamma_user,
+        'Gi': gamma_item, 'Tu': theta_user, 'E': emb_matrix, 'Bp': beta_prime}
+
+    seed: int, optional, default: None
+        Random seed for weight initialization.
+
+    References
+    ----------
+    * He, R., & McAuley, J. (2016). VBPR: Visual Bayesian Personalized Ranking from Implicit Feedback.
+    """
+    
+    def __init__(
+            self,
+            name="AMR",
+            k=10,
+            k2=10,
+            n_epochs=50,
+            batch_size=100,
+            learning_rate=0.005,
+            lambda_w=0.01,
+            lambda_b=0.01,
+            lambda_e=0.0,
+            lmd=1.0,
```

Comment:
I would suggest to rename `lmd` to `lambda_adv` for readability. It's also more consistent with other regularization hyper-params.